### ISSUE TITLE
cabal list --installed does not list installed packages

### ISSUE THREAD

**Issue by JonasRSV** [haskell/cabal#7069]
**Describe the bug**
I installed xmonad with

cabal install xmonad

but when I run cabal list --installed xmonad does not show up.

**To Reproduce**

Download precompiled cabal 3.2.0.0 binary from https://www.haskell.org/cabal/download.html for x86-64
Installed ghc from https://www.haskell.org/ghc/download_ghc_8_6_5.html

Then run cabal install xmonad 
then cabal list --installed

xmonad does not show up.

**Expected behavior**
Xmonad shows up.

**System information**
 - Ubuntu 18.0.4.3
 - ghc version 8.6.5
 -  cabal verison 3.2.0.0

**Additional context**

My real problem is that when I am trying to compile xmonad with ghc-8.6.5 it can't resolve the xmonad lib.

I used to have ghc-8.0.1 and I have a few times tried to upgrade but everything keeps breaking and I usually just give up after a few hours.. this time I am not giving up so easily though. 

It appears as if ghc 8.6.5 is bundled cabal 2.4 with it and ghc-pkg seems to be dependant on it. I have tried to find out how I can make ghc-pkg use the new cabal version but to no avail. 

My guess is that cabal list --installed is using the ghc-pkg command which in turn uses the wrong cabal program? I don't know though. How do I make my cabal installs available to my ghc 8.6.5 installation?

**Comment by fgaz** [comment]
Basically `cabal list --installed` lists the _globally_ installed packages, while recently cabal switched to a [new model](https://cabal.readthedocs.io/en/3.4/nix-local-build-overview.html) where packages are not installed globally anymore (it would become unmanageable at some point) and are instead isolated from each other

> I am trying to compile xmonad with ghc-8.6.5 

I guess you are invoking ghc directly?
My suggestion would be to create a cabal project for your xmonad configuration and to add the xmonad library to the build-depends section, so that the dependency is managed declaratively.
Alternatively, you could use something like `cabal install --package-env=xmonad-env --lib the-xmonad-library-you-are-trying-to-use` to create a package environment called `xmonad-env` containing the xmonad library, so that you can tell ghc to use it like this `ghc -package-env xmonad-env Somefile.hs`

**Comment by JonasRSV** [comment]
not sure, I am using 'xmonad --recompile' I believe xmonad is invoking ghc. Not sure how to make xmonad use local packages.

So I can't install packages globally anymore? What version of cabal would I have to download to get that capability back?

**Comment by fgaz** [comment]
You can also use [the GHC_ENVIRONMENT environment variable](https://downloads.haskell.org/ghc/latest/docs/html/users_guide/packages.html#envvar-GHC_ENVIRONMENT) to do the same.

Or, (discouraged), you can install libraries in the default environment, which as the name implies is picked up by default by ghc, by omitting the --package-env flag in the cabal invocation

**Comment by fgaz** [comment]
And yes, xmonad invokes ghc, so the first method should work

**Comment by JonasRSV** [comment]
How would I pass the package-env flag to ghc through xmonad though?

How would I install things in the default environment? I really would prefer not creating some package environment.. I just want the dependencies for my window manager globally accessible. I guess installing into the default environment would do that? So how do I install there? Because when I run 

```bash
cabal install xxx
```

it installs into someplace else that is not found by ghc.

**Comment by JonasRSV** [comment]
I would love to just be able to install packages into my global package db which is: /usr/lib/ghc-8.6.5/package.conf.d since that is what my ghc-pkg list is listing.

**Comment by JonasRSV** [comment]
So I just recompiled in a local environment and now my window manager is broken, I cannot log in using xmonad anymore.. Is there any differences between compiling in a local environment and global that can help me debug this?

**Comment by fgaz** [comment]
> How would I install things in the default environment?

libraries are not installed by default, since that often isn't what the user wants, so you have to use the --lib flag:

    cabal install --lib thelibrary

> I would love to just be able to install packages into my global package db which is: /usr/lib/ghc-8.6.5/package.conf.d since that is what my ghc-pkg list is listing.

If you want to use the old behavior, `cabal v1-install thelibrary` should still work

**Comment by phadej** [comment]
I suggest that you hold this discussion, and someone who

- knows `xmonad`
- knows `cabal v2-build`

writes a tutorial on how to combine this. My gut feeling is that `xmonad` have to be updated to support `cabal v2-build` workflows, (e.g. call `ghc` with an environment parameters). Yet even better would be to have `xmonad.hs` (the configuration file) have `build-depends: ..` specification which somehow would be picked by `cabal`. We have something like that for script execution, but here we want to compile an executable.

**TL;DR** to make `xmonad` and `cabal v2-build` work together, we need changes to both.

If you are not intimately familiar with both, I wouldn't recommend using both. (i.e. pick one).

**Comment by fgaz** [comment]
By the way to me it looks like what would work best for you would probably be a xmonad system package. Doesn't your distribution package it? Also I just looked at the [xmonad guide](https://xmonad.org/intro.html) and it looks quite outdated: it mentions a cabal release from four years ago! Maybe you could bring this up to the maintainers

**Comment by phadej** [comment]
> Also I just looked at the xmonad guide and it looks quite outdated: it

Yes, that is my impression as well. `xmonad` is "simply" not updated to work with `v2-build` paradigm. It's very well possible, that some small bit is also missing from `cabal`, but we don't know which (we have an idea, see above comment).

**Comment by fgaz** [comment]
I'm pretty sure we had this same discussion with another xmonad user some time ago, and in the end they build a v2-based workflow that worked well for them. I think they also had the intention of writing a guide, but I'm not sure what came of it

**Comment by JonasRSV** [comment]
Where can I find the cabal-install version 1.24? On this page it seems that only the library is https://www.haskell.org/cabal/download.html is available.

I am just trying to get my old stuff working again, and in my 
stupidity I removed my old cabal installation when installing the new one.

I originally installed xmonad using ghc 8.0.1 and cabal 1.24, I just thought it'd be nice to update ghc and cabal. but everything just seems to break, so i'll stick with my old ghc and cabal.. if that cabal version can be downloaded again from somewhere :)

**Comment by JonasRSV** [comment]
I found it here https://hackage.haskell.org/package/cabal-install-1.24.0.2

**Comment by JonasRSV** [comment]
Will close this issue as soon as I got my old things working again.

**Comment by JonasRSV** [comment]
> I found it here https://hackage.haskell.org/package/cabal-install-1.24.0.2

Nvm, these  versions don't build with ghc 8.0.2 or ghc 8.6.5... is there any precompiled binaries available? :(((((

**Comment by JonasRSV** [comment]
Apparently I cannot install cabal 1.24 again because building it requires Data.Hashtable which was deprecated in Base 4.7 and I have Base 4.9.  And changing to Base 4.7 Would break a bunch of other packages. Wonderful.

**Comment by JonasRSV** [comment]
Well the solution to getting back the WM was to just install it manually using the Setup.hs files from 

https://hackage.haskell.org/package/xmonad
